colcon-ros-cargo/colcon_ros_cargo/task/ament_cargo/build.py

```diff
@@ -0,0 +1,180 @@
+
+from colcon_core.logging import colcon_logger
+from colcon_core.plugin_system import satisfies_version
+from colcon_core.shell import create_environment_hook
+from colcon_core.task import TaskExtensionPoint
+from colcon_core.task import run
+
+import toml
+
+
+logger = colcon_logger.getChild(__name__)
+
+# Some logic needs to be executed once per run.
+# There are no colcon hooks for this, so it is shoehorned into the build step
+# with a global.
+package_paths = None
+
+
+class AmentCargoBuildTask(CargoBuildTask):
+    """A build task for packages with Cargo.toml + package.xml.
+
+    The primary problem that needs to be solved is that dependencies on other
+    packages in the same workspace are expressed by just a name in colcon, but
+    by a full path in Cargo.
+
+    That means, when building a Cargo package, all the packages it depends on
+    need to be resolved to full paths, and those need to be written into the
+    Cargo.toml, or alternatively into a [patch] section of a .cargo/config.toml
+    file. Here the latter approach is used.
+    """
+
+    def __init__(self):  # noqa: D107
+        super().__init__()
+        satisfies_version(TaskExtensionPoint.EXTENSION_POINT_VERSION, '^1.0')
+
+    def add_arguments(self, *, parser):  # noqa: D102
+        parser.add_argument(
+            '--lookup-only-install',
+            action='store_true',
+            help='Look up dependencies in the installation prefixes only. '
+            'By default, dependencies are looked up in the workspace whenever '
+            "possible, which does not strictly follow colcon's intended "
+            'semantics.')
+
+    async def build(  # noqa: D102
+        self, *, additional_hooks=[], skip_hook_creation=False
+    ):
+        additional_hooks += create_environment_hook(
+            'ament_prefix_path',
+            Path(self.context.args.install_base),
+            self.context.pkg.name,
+            'AMENT_PREFIX_PATH',
+            self.context.args.install_base,
+            mode='prepend')
+
+        return await super(AmentCargoBuildTask, self).build(
+            additional_hooks=additional_hooks,
+            skip_hook_creation=skip_hook_creation
+        )
+
+    async def _build(self, args, env):
+        self.progress('prepare')
+
+        global package_paths
+        if package_paths is None:
+            if args.lookup_only_install:
+                package_paths = {}
+            else:
+                package_paths = find_workspace_cargo_packages(args.install_base)  # noqa: E501
+
+        # Any message Rust packages that were generated after the build process
+        # started are not yet in package_paths. So we'll need to scan the
+        # install dirs, aka prefixes.
+        new_package_paths = find_installed_cargo_packages(env)
+        new_package_paths.update(package_paths)
+        package_paths = new_package_paths
+        write_cargo_config_toml(package_paths)
+
+        # Clean up the build dir
```

Comment:
Cleaning the build dir makes incremental compilation impossible. Should we add a flag to not clean? Or even make incremental the default and add a --clean flag?

Reply:
The problem with not cleaning is that if you remove a file from the source space, how would that be reflected in the build space? In that case, the `colcon-ros-cargo` would need to do something like `rsync` to keep both trees in sync. It's doable, but it's up to you if you to implement something like that. For now I think it's ok to clean the build space and improve the logic there later.

Reply:
Well, the build dir = cargo's target dir. Cargo is usually very good at not picking up outdated build artifacts (using hashes and whatnot). It's just in rare cases that a user might need to delete the build/target dir. So I've now made it the default to not delete it and added a `--clean-build` flag.